hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1360,7 +1361,10 @@ protected void compactIfNecessary(BaseHoodieWriteClient writeClient, String late
     // The compaction planner will manage to filter out the log files that finished with greater completion time.
     // see BaseHoodieCompactionPlanGenerator.generateCompactionPlan for more details.
     final String compactionInstantTime = dataMetaClient.reloadActiveTimeline().filterInflightsAndRequested()
-        .findInstantsBeforeOrEquals(latestDeltacommitTime).firstInstant().map(HoodieInstant::getTimestamp)
+        .findInstantsBeforeOrEquals(latestDeltacommitTime).firstInstant()
+        // minus the pending instant time by 1 millisecond to avoid conflict in case when this pending instant was finally been committed
+        // as a delta_commit in MDT.
+        .map(instant -> HoodieInstantTimeGenerator.instantTimeMinusMillis(instant.getTimestamp(), 1L))
```

Comment:
Based on the logic, previously the compaction would never be scheduled, given it picks the instant time that already exists (entering the if branch logging `Compaction with same %s time is already present in the timeline.`).

Reply:
> Compaction with same %s time is already present in the timeline.).

Not really, the earlist pending instant on DT may not have a committed delta_commit on MDT timeline.